app/code/Magento/MediaGallery/Model/ResourceModel/Keyword/SaveAssetLinks.php

```diff
@@ -56,27 +66,124 @@ public function __construct(
     public function execute(int $assetId, array $keywordIds): void
     {
         try {
-            $values = [];
-            foreach ($keywordIds as $keywordId) {
-                $values[] = [$assetId, $keywordId];
+            $currentKeywordIds = $this->getCurrentKeywordIds($assetId);
+
+            $obsoleteKeywordIds = array_diff($currentKeywordIds, $keywordIds);
+            $newKeywordIds = array_diff($keywordIds, $currentKeywordIds);
+
+            $this->deleteAssetKeywords($assetId, $obsoleteKeywordIds);
+            $this->insertAssetKeywords($assetId, $newKeywordIds);
+
+        } catch (\Exception $exception) {
+            $this->logger->critical($exception);
+            throw new CouldNotSaveException(
+                __('Could not process asset keyword links'),
+                $exception
+            );
+        }
+    }
+
+    /**
+     * Save new asset keyword links
+     *
+     * @param int $assetId
+     * @param int[] $keywordIds
+     * @throws CouldNotSaveException
+     */
+    private function insertAssetKeywords(int $assetId, array $keywordIds): void
+    {
+        try {
+            if (!empty($keywordIds)) {
+                $values = [];
+
+                foreach ($keywordIds as $keywordId) {
+                    $values[] = [$assetId, $keywordId];
+                }
+
+                if (!empty($values)) {
```

Comment:
I think this is an unnecessary check. the previous check should be enough in my opinion